juju/model.py

```diff
@@ -0,0 +1,210 @@
+import json
+from collections.abc import Mapping
+from functools import lru_cache
+from subprocess import run, PIPE
+
+
+class Model:
+    def __init__(self, app_name, unit_name, relation_names, backend=None):
+        self.app = Application(app_name, self._lazy_load_peer_units)
+        self.unit = Unit(unit_name)
+        self._backend = backend or ModelBackend()
+        self.relations = RelationMap(relation_names, self.app, self.unit, self._backend)
+
+    def relation(self, relation_name):
+        """Return the first Relation object for the named relation, or None.
+
+        This is a convenience method for the case where only a single related
+        app is supported. If there are more than one remote apps on the relation,
+        it will raise a TooManyRelatedApps model error. If no there are no remote
+        apps on the relation, it will return None. Otherwise, it is equivalent
+        to self.relations[relation_name][0].
+        """
+        num_related = len(self.relations[relation_name])
+        if num_related > 1:
+            raise TooManyRelatedApps(relation_name, num_related, 1)
+        elif num_related == 0:
+            return None
+        else:
+            return self.relations[relation_name][0]
+
+    def _lazy_load_peer_units(self):
+        # In lieu of an implicit peer relation, goal-state seems to be the only way to find out
+        # what our peer units are. The information it returns is rather limited, but is enough
+        # for our purposes here.
+        return [Unit(unit_name) for unit_name in self._backend.goal_state()['units']]
```

Comment:
It doesn't seem to make sense to have goal state here. We cannot act on those units in a normal way before they actually join the relationship. They might not even exist at all right now. Per conversations in the sprint, I think we'll want to rethink the problem addressed by goal state before we use this.

This is also misplaced. More details on that below.

Reply:
I mentioned in the comment that currently, `goal-state` is the only way to get the list of peer units, currently.  If we can force or depend on an implicit peer relation, then I'm all for assuming that an `Application` is always associated with a relation ID and moving the `Unit` loading logic to that.  However, I should also note that `relation-list` only ever lists remote units, so there will still need to be some way to inform and handle the case where the `Application` should include the local unit.